File: frontend/src/types/operations.ts
export interface OperationResult {
  success: boolean
  message: string
  task_id?: string
  details: Record<string, unknown>
}

export interface OperationsOverview {
  timestamp: string
  nginx_available: boolean
  worker_queue_depth: number
  backup_count: number
  cron_job_count: number
  applications_total: number
  applications_enabled: number
}

export interface EnvVariable {
  key: string
  value: string
  secret: boolean
  source?: string
}

export interface EnvironmentResponse {
  timestamp: string
  variables: EnvVariable[]
  revealed: boolean
}

export interface FileEntry {
  name: string
  path: string
  is_dir: boolean
  size_bytes?: number
  modified?: string
  mode?: string
  owner?: string
  group?: string
}

export interface FileListResponse {
  timestamp: string
  path: string
  entries: FileEntry[]
  parent?: string
}

export interface StorageVolume {
  mount: string
  total_bytes: number
  used_bytes: number
  free_bytes: number
  percent: number
}

export interface BackupEntry {
  id: string
  name: string
  path: string
  size_bytes?: number
  created?: string
  application_id?: string
}

export interface CronJob {
  id: string
  schedule: string
  command: string
  source: string
  user?: string
  enabled: boolean
}

export interface DatabaseStatus {
  engine: string
  status: string
  message?: string
  size_bytes?: number
  connections?: number
}

export interface SslAppStatus {
  application_id: string
  application_name: string
  domain?: string
  ssl: Record<string, unknown>
}

export interface ApplicationDetail {
  id: string
  name: string
  type: string
  environment: string
  enabled: boolean
  status: string
  health: string
  health_score: number
  domain?: string
  root_path?: string
  version?: string
  description?: string
  tags: string[]
  paths: Record<string, unknown>
  runtime: Record<string, unknown>
  git?: Record<string, unknown>
  ssl?: Record<string, unknown>
  nginx?: Record<string, unknown>
  supervisor?: Record<string, unknown>
  systemd?: Record<string, unknown>
  modules: Record<string, boolean>
}

export interface ApplicationLogsResponse {
  timestamp: string
  application_id: string
  sources: string[]
  entries: Array<{ message?: string; level?: string; source?: string; line_number?: number }>
}

export interface DeploymentRecord {
  id: string
  version: string
  status: string
  environment: string
  timestamp: string
  triggered_by?: string
  message?: string
}

File: backend/app/schemas/operations.py
"""Operations control plane schemas."""

from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.common import SchemaBase
from app.schemas.health import HealthStatus


class OperationResult(SchemaBase):
    """Result of a mutating operation."""

    success: bool
    message: str
    task_id: str | None = None
    details: dict[str, Any] = Field(default_factory=dict)


class EnvVariable(SchemaBase):
    key: str
    value: str
    secret: bool = False
    source: str | None = None


class EnvironmentResponse(SchemaBase):
    timestamp: datetime
    variables: list[EnvVariable]
    revealed: bool = False


class SmtpTestRequest(SchemaBase):
    to_email: str
    subject: str = "IFNOTUS SMTP Test"
    body: str = "This is a test message from IFNOTUS."


class FileEntry(SchemaBase):
    name: str
    path: str
    is_dir: bool
    size_bytes: int | None = None
    modified: datetime | None = None
    mode: str | None = None
    owner: str | None = None
    group: str | None = None


class FileListResponse(SchemaBase):
    timestamp: datetime
    path: str
    entries: list[FileEntry]
    parent: str | None = None


class StorageVolume(SchemaBase):
    mount: str
    total_bytes: int
    used_bytes: int
    free_bytes: int
    percent: float


class StorageResponse(SchemaBase):
    timestamp: datetime
    volumes: list[StorageVolume]


class BackupEntry(SchemaBase):
    id: str
    name: str
    path: str
    size_bytes: int | None = None
    created: datetime | None = None
    application_id: str | None = None


class BackupsResponse(SchemaBase):
    timestamp: datetime
    backups: list[BackupEntry]


class CronJob(SchemaBase):
    id: str
    schedule: str
    command: str
    source: str
    user: str | None = None
    enabled: bool = True


class CronResponse(SchemaBase):
    timestamp: datetime
    jobs: list[CronJob]


class DatabaseStatus(SchemaBase):
    engine: str
    status: HealthStatus
    message: str | None = None
    size_bytes: int | None = None
    connections: int | None = None


class DatabaseResponse(SchemaBase):
    timestamp: datetime
    databases: list[DatabaseStatus]


class ServiceControlRequest(SchemaBase):
    service: str
    source: str = "systemd"  # systemd | supervisor | nginx


class QueueStatus(SchemaBase):
    queue: str
    depth: int
    worker_running: bool | None = None


class OperationsOverview(SchemaBase):
    timestamp: datetime
    nginx_available: bool
    worker_queue_depth: int
    backup_count: int
    cron_job_count: int
    applications_total: int
    applications_enabled: int
